lib/uplink/project.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"context"
+
+	"github.com/vivint/infectious"
+
+	"storj.io/storj/internal/memory"
+	"storj.io/storj/pkg/eestream"
+	"storj.io/storj/pkg/encryption"
+	"storj.io/storj/pkg/metainfo/kvmetainfo"
+	"storj.io/storj/pkg/storage/buckets"
+	ecclient "storj.io/storj/pkg/storage/ec"
+	"storj.io/storj/pkg/storage/segments"
+	"storj.io/storj/pkg/storage/streams"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/pkg/transport"
+	"storj.io/storj/uplink/metainfo"
+)
+
+// Project represents a specific project access session.
+type Project struct {
+	tc       transport.Client
+	metainfo metainfo.Client
+	project  *kvmetainfo.Project
+}
+
+// CreateBucketOptions holds possible options that can be passed to
+// CreateBucket.
+type CreateBucketOptions struct {
+	PathCipher Cipher
+}
+
+func (o *CreateBucketOptions) setDefaults() {
+	if o.PathCipher == UnsetCipher {
+		o.PathCipher = defaultCipher
+	}
+}
+
+// CreateBucket creates a new bucket if authorized
+func (p *Project) CreateBucket(ctx context.Context, bucket string, opts CreateBucketOptions) (b storj.Bucket, err error) {
+	defer mon.Task()(&ctx)(&err)
+	opts.setDefaults()
+	pathCipher, err := opts.PathCipher.convert()
+	if err != nil {
+		return storj.Bucket{}, err
+	}
+	return p.project.CreateBucket(ctx, bucket, &storj.Bucket{PathCipher: pathCipher})
+}
+
+// DeleteBucket deletes a bucket if authorized
+func (p *Project) DeleteBucket(ctx context.Context, bucket string) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	return p.project.DeleteBucket(ctx, bucket)
+}
+
+// ListBuckets will list authorized buckets
+func (p *Project) ListBuckets(ctx context.Context, opts storj.BucketListOptions) (bl storj.BucketList, err error) {
+	defer mon.Task()(&ctx)(&err)
+	return p.project.ListBuckets(ctx, opts)
+}
+
+// GetBucketInfo returns info about the requested bucket if authorized
+func (p *Project) GetBucketInfo(ctx context.Context, bucket string) (b storj.Bucket, err error) {
+	defer mon.Task()(&ctx)(&err)
+	return p.project.GetBucket(ctx, bucket)
+}
+
+// BucketConfig represents configuration options for a specific Bucket
+type BucketConfig struct {
+	EncryptionAccess EncryptionAccess
+
+	// These config values are likely to change semantics or go away
+	// entirely between releases. Be careful when using them!
+	Volatile struct {
+		DefaultRS struct {
+			MinThreshold     int
+			RepairThreshold  int
+			SuccessThreshold int
+			MaxThreshold     int
+		}
+
+		MaxBufferMem     memory.Size
+		ErasureShareSize memory.Size
+		SegmentSize      memory.Size
+		MaxInlineSize    memory.Size
```

Comment:
This one surely doesn't belong here, does it? This seems like it should be a decision made by the satellite operator and no one else.

Reply:
oh, the satellite operator definitely decides whether to accept or reject, but this is the hint on the client side of what to do, since the satellite doesn't actually advertise it's setting anywhere. this is in volatile though, right?

Reply:
Okay. Yeah, it is, I'm just trying to sanity check